cdap-tms/src/main/java/io/cdap/cdap/messaging/store/leveldb/LevelDBPartitionManager.java

```diff
@@ -0,0 +1,305 @@
+    return partitionStartTime == null ? partitions.values() : partitions.tailMap(partitionStartTime, true).values();
+  }
+
+  /**
+   * Get the partitions responsible for data between the specified start and end times.
+   *
+   * @param startTime inclusive start time
+   * @param endTime inclusive end time
+   * @return partitions responsible for data between the specified start and end times
+   */
+  public Collection<LevelDBPartition> getPartitions(long startTime, long endTime) throws IOException {
+    if (initialized.compareAndSet(false, true)) {
+      initPartitions();
+    }
+    Long partitionEndTime = partitions.floorKey(endTime);
+    if (partitionEndTime == null) {
+      // partitionEndTime can only be null if the partitions map is empty, or the endTime is less than the
+      // start time for any existing partition. In either case, there are no partitions that contain data
+      // published between the start and end times.
+      return Collections.emptyList();
+    }
+
+    Long partitionStartTime = partitions.floorKey(startTime);
+    // partitionStartTime can be null if the startTime is less than the start time for any existing partition
+    partitionStartTime = partitionStartTime == null ? 0 : partitionStartTime;
+    return partitions.subMap(partitionStartTime, true, partitionEndTime, true).values();
+  }
+
+  public void close() throws IOException {
+    List<LevelDBPartition> partitionsToClose;
+
+    synchronized (this) {
+      partitionsToClose = new ArrayList<>(partitions.values());
+      partitions.clear();
+    }
+
+    IOException failure = null;
+    for (LevelDBPartition partition : partitionsToClose) {
+      try {
+        partition.close();
+      } catch (IOException e) {
+        if (failure == null) {
+          failure = e;
+        } else {
+          failure.addSuppressed(e);
+        }
+      }
+    }
+    if (failure != null) {
+      throw failure;
+    }
+    partitions.clear();
+  }
+
+  @Nullable
+  private LevelDBPartition getPartitionIfExists(long publishTime) {
+    Long partitionKey = partitions.floorKey(publishTime);
+    if (partitionKey != null) {
+      LevelDBPartition partition = partitions.get(partitionKey);
+      if (partition.getEndTime() > publishTime) {
+        return partition;
+      }
+    }
+    return null;
+  }
+
+  private synchronized void initPartitions() throws IOException {
+    ensureDirExists(topicDir);
+    File[] partitionDirs = topicDir.listFiles();
+
+    if (partitionDirs != null) {
+      for (File partitionDir : partitionDirs) {
+        if (!partitionDir.isDirectory()) {
+          continue;
+        }
+
+        String dirName = partitionDir.getName();
+        TimeInterval interval = extractIntervalFromName(dirName);
+        if (interval == null) {
+          // should not happen unless somebody manually created a directory here
+          continue;
+        }
+
+        DB db = LEVEL_DB_FACTORY.open(partitionDir, dbOptions);
```

Comment:
It might be better to not open the DB, but only open it when it is needed. You could wrap with a supplier to delay the opening.